eos-ostree-remotes-config

```diff
@@ -28,16 +28,119 @@
+REPO_OBJECTS_PATH = os.path.join(REPO_PATH, 'objects')
+OSTREE_REMOTES_PATH = '/etc/ostree/remotes.d'
+FLATPAK_REMOTES_PATH = '/etc/flatpak/remotes.d'
+
+
+def read_config_file(config, path):
+    """Read the file at path into config
+
+    Returns False if the file doesn't exist or could not be parsed.
+    """
+    try:
+        print('Parsing', path)
+        with open(path) as f:
+            config.read_file(f)
+        return True
+    except FileNotFoundError as err:
+        print('Warning:', path, 'does not exist', file=sys.stderr)
+        print(err, file=sys.stderr)
+        return False
+    except ParsingError as err:
+        print('Warning: could not parse', path, file=sys.stderr)
+        print(err, file=sys.stderr)
+        return False
+
+
+aparser = ArgumentParser(
+    description='Configure /ostree/repo for safe remote adding')
+aparser.add_argument('-n', '--dry-run', action='store_true',
+                     help='only show what would be done')
+args = aparser.parse_args()
+
 
 # Check that /ostree/repo looks like a real repo
-[ -d /ostree/repo ] || exit 0
-[ -f /ostree/repo/config ] || exit 0
-[ -d /ostree/repo/objects ] || exit 0
-
-# Only set the option if it hasn't been set yet
-CONFIG_OPT="core.add-remotes-config-dir"
-if ! ostree --repo=/ostree/repo config get "$CONFIG_OPT" &>/dev/null; then
-    ostree --repo=/ostree/repo config set "$CONFIG_OPT" false
-fi
+if (not os.path.isdir(REPO_PATH) or
+    not os.path.isfile(REPO_CONFIG_PATH) or
+    not os.path.isdir(REPO_OBJECTS_PATH)):
+    sys.exit(0)
+
+# Read repo config
+repo_config = ConfigParser()
+if not read_config_file(repo_config, REPO_CONFIG_PATH):
+    # This is bad...
+    print('Error: could not read', REPO_CONFIG_PATH, file=sys.stderr)
+    sys.exit(1)
+
+# Set the core.add-remotes-config-dir option and write back now in case
+# there are errors later
+if repo_config.has_option('core', REMOTE_CONFIG_OPT):
+    value = repo_config.get('core', REMOTE_CONFIG_OPT)
+    if value != 'false':
+        print('Warning: unsupported {} core.{} value {} in {}'
+              .format(REMOTE_CONFIG_OPT, value, REPO_CONFIG_PATH))
+else:
+    if not repo_config.has_section('core'):
+        repo_config.add_section('core')
+    print('Setting option core.{}=false'.format(REMOTE_CONFIG_OPT))
+    repo_config.set('core', REMOTE_CONFIG_OPT, 'false')
+    if not args.dry_run:
+        print('Writing', REPO_CONFIG_PATH)
+        with open(REPO_CONFIG_PATH, 'w') as f:
+            repo_config.write(f, space_around_delimiters=False)
+
+# Read all the ostree and flatpak remotes.d files into the
+# configuration. This has the effect of merging them into the repo
+# config.
+remotes_d_confs = []
+for remotes_dir in (OSTREE_REMOTES_PATH, FLATPAK_REMOTES_PATH):
+    print('Reading config files in', remotes_dir)
+    for path in glob.iglob(remotes_dir + '/*.conf'):
+        remotes_d_confs.append(path)
+        read_config_file(repo_config, path)
```

Comment:
Simply merging in the remotes.d entries isn't sufficient to leave the user with a working remote setup if flatpak remote-add failed mid-flow due to the remotes.d error:
```root@endless:~# cat /etc/flatpak/remotes.d/flathub.conf 
[remote "flathub"]
url=https://dl.flathub.org/repo/```
(and no GPG key!)

Its existence will prevent eos-update-flatpak-repos trying to re-add it. We should either wipe them out aggressively, safe in the knowledge (?!) that eos/eos-apps have always been happily defined in config, or we need to make eos-update-flatpak-repos sense check the remotes it would add, and remove --force then re-add them if they are missing eg URL or GPG key. (see 48b1f06c8b2ae4cf09076e252b80af73318f3a5c for prior art)

Reply:
Good point. I wanted to keep any legit user defined entries there, but I agree that the common case is broken remotes handled by `eos-update-flatpak-repos`.

So, I think the thing to do is keep a list of remotes that will be added by `eos-update-flatpak-repos`. If we find any of those in one of the remotes.d dirs, just get rid of it entirely with the knowledge that `eos-update-flatpak-repos` will bring it back correctly.